go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,271 @@
+package clustermanager
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	nad "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/network-attach-def-controller"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// ovnkubeClusterManagerLeaderMetrics object is used for the cluster manager
+// leader metrics
+type ovnkubeClusterManagerLeaderMetrics struct{}
+
+func (ovnkubeClusterManagerLeaderMetrics) On(string) {
+	metrics.MetricClusterManagerLeader.Set(1)
+}
+
+func (ovnkubeClusterManagerLeaderMetrics) Off(string) {
+	metrics.MetricClusterManagerLeader.Set(0)
+}
+
+type ovnkubeClusterManagerLeaderMetricsProvider struct{}
+
+func (_ ovnkubeClusterManagerLeaderMetricsProvider) NewLeaderMetric() leaderelection.SwitchMetric {
+	return ovnkubeClusterManagerLeaderMetrics{}
+}
+
+// ClusterManager structure is the object which manages the cluster nodes.
+// It creates a default network controller for the default network and a
+// secondary network cluster controller manager to manage the multi networks.
+type ClusterManager struct {
+	client                               clientset.Interface
+	defaultNetClusterController          *defaultNetworkClusterController
+	wf                                   *factory.WatchFactory
+	wg                                   *sync.WaitGroup
+	stopChan                             chan struct{}
+	secondaryNetClusterControllerManager *secondaryNetworkClusterControllerManager
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+}
+
+// NewClusterManager creates a new Cluster Manager for managing the
+// cluster nodes.
+func NewClusterManager(ovnClient *util.OVNClusterManagerClientset, wf *factory.WatchFactory,
```

Comment:
I think you should pass the wait group plus stop chan in `Start`.

Reply:
Basically I'm following how the existing code creates network controller manager.

https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/cmd/ovnkube/ovnkube.go#L252

And the function NewNetworkControllerManager() here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/network-controller-manager/network_controller_manager.go#L278  creates its own stopChannel and this is used for stopping  the secondary network controllers.

And the default network controller is creating its own stop channel - https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/default_network_controller.go#L151


I'm confused whats the correct way.  I don't have any strong opinion.  I just thought I'd follow the same design pattern.

Reply:
Talked to Numan. We can rework  this in the future to make it consistent.

Reply:
+1
| Also we could probably just work with the context provided and not need any stop chans or stop methods.